src/private/runApi.ts

```diff
@@ -0,0 +1,145 @@
+  ActionOptions,
+  CroodsState,
+  HeadersObj,
+  HttpMethod,
+  ReqBody,
+  SaveOptions,
+} from 'types'
+import type { ObjWithStore } from '../useStore'
+
+type RequestType =
+  | { kind: 'fetch'; piece: CroodsState }
+  | { kind: 'destroy' }
+  | { kind: 'save'; options: SaveOptions }
+
+type Request = {
+  actions: ObjWithStore<InternalActions>
+  options: ActionOptions
+  requestType: RequestType
+}
+
+const getMethod = (options: ActionOptions, type: RequestType): HttpMethod => {
+  if (typeof options.method === 'string') return options.method
+  switch (type.kind) {
+    case 'fetch':
+      return 'GET'
+    case 'destroy':
+      return 'DELETE'
+    case 'save':
+      return options.id ? 'PUT' : 'POST'
+  }
+}
+
+const getHeaders = async (options: ActionOptions): Promise<HeadersObj> => {
+  const defaultHeaders: HeadersObj = {
+    Accept: 'application/json',
+    'Content-Type': 'application/json',
+  }
+  return typeof options.headers === 'function'
+    ? options.headers(defaultHeaders)
+    : { ...defaultHeaders, ...options.headers }
+}
+
+const getUrl = (options: ActionOptions): string => {
+  const path = buildUrl(options)
+  const queryString = buildQueryString(options)
+  return joinWith('?', path, queryString)
+}
+
+const shouldUseCache = ({ options, requestType }: Request): boolean => {
+  return Boolean(
+    requestType.kind === 'fetch' &&
+      Boolean(requestType.piece?.list?.length) &&
+      options.cache,
+  )
+}
+
+const requestAction = ({ actions, options, requestType }: Request): boolean => {
+  switch (requestType.kind) {
+    case 'fetch':
+      return actions.getRequest(options)
+    case 'save':
+      return actions.saveRequest(options)
+    case 'destroy':
+      return actions.destroyRequest(options)
+  }
+}
+
+const successAction = <T = any>(
+  { actions, options, requestType }: Request,
+  result: T,
+): T => {
+  switch (requestType.kind) {
+    case 'fetch':
+      return actions.getSuccess(options, result)
+    case 'save':
+      return actions.saveSuccess(options, result)
+    case 'destroy':
+      return actions.destroySuccess(options)
+  }
+}
+
+const failureAction = (
+  { actions, options, requestType }: Request,
+  error: string,
+): false => {
+  switch (requestType.kind) {
+    case 'fetch':
+      return actions.getFail(options, error)
+    case 'save':
+      return actions.saveFail(options, error)
+    case 'destroy':
+      return actions.destroyFail(options, error)
+  }
+}
+
+async function runApi<T>(
+  request: Request,
+): Promise<(body?: ReqBody) => Promise<T>> {
+  const { actions, options, requestType } = request
+  const url = getUrl(options)
+  const headers = await getHeaders(options)
+  const method = getMethod(options, requestType)
+  return async (body?: ReqBody) => {
+    const config = { method, headers } as RequestInit
+    if (body) {
+      const parseParams = createHumps(options.paramsParser || snakeCase)
+      const data = parseParams(omit(body, 'id'))
```

Comment:
Is this removing `id` from the body? why?

Reply:
This is a decision took in the old version.. can't recall exactly why... if we can't think of a good reason let's remove this behavior given this PR already introduces breaking changes